### PR TITLE
improve provider specific options

### PR DESCRIPTION
**Pull request by MayNiklas**
* I've added a new host for our coming attic server (running on Proxmox)
* I've added a Proxmox module
* I made Nginx optional (attic module)
* I combined all the provider specific modules into one module: `cloud-provider`

The last one is quite important in my opinion:
Before that, 50% of our modules would be provider modules. Now they are all within one folder combined into one module. This makes it easier to navigate the files. Also: imports (for example disko) are more straight forward.

### CODE REVIEW

File: machines/attic-server/configuration.nix [hunk 1]
```diff
@@ -0,0 +1,23 @@
+{ self, ... }:
+{ pkgs, lib, config, ... }:
+{
+
+  networking = {
+    hostName = "attic-server";
+  };
```
Comment (pinpox) [inline machines/attic-server/configuration.nix:7]:
```suggestion
  networking.hostName = "attic-server";
```

File: machines/attic-server/configuration.nix [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{ self, ... }:
+{ pkgs, lib, config, ... }:
+{
+
```
Comment (pinpox) [inline machines/attic-server/configuration.nix:4]:
```suggestion
{ pkgs, lib, config, ... }:
{

```

File: machines/attic-server/configuration.nix [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  };
+
+  mayniklas.user.root.enable = true;
+
+  lounge-rocks = {
+    # attic.server.enable = true;
+    nix-common.enable = true;
+    cloud-provider = {
+      enable = true;
+      proxmox.enable = true;
```
Comment (pinpox) [inline machines/attic-server/configuration.nix:16]: Is a separate `enable` option really necessary? having just `proxxmox.enable` etc. would make sense

File: machines/attic-server/configuration.nix [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    };
+  };
+
+  system.stateVersion = "23.05";
```
Comment (pinpox) [inline machines/attic-server/configuration.nix:20]: We might want to put the `stateVersion` into a commonly imported module

File: machines/netcup-x86-runner-1/configuration.nix [hunk 5]
```diff
@@ -2,7 +2,10 @@
 { pkgs, lib, ... }: {
 
   lounge-rocks = {
-    netcup-x86.enable = true;
+    cloud-provider = {
+      enable = true;
+      netcup.enable = true;
+    };
```
Comment (pinpox) [inline machines/netcup-x86-runner-1/configuration.nix:8]: Same as above. Consider dropping  `cloud-provider.enable` in favor of just having `cloud-provider.netcup.enable`

File: machines/oracle-aarch64-runner-1/configuration.nix [hunk 6]
```diff
@@ -2,7 +2,10 @@
 { pkgs, lib, ... }: {
 
   lounge-rocks = {
-    oracle-aarch64.enable = true;
+    cloud-provider = {
+      enable = true;
+      oracle.enable = true;
+    };
```
Comment (pinpox) [inline machines/oracle-aarch64-runner-1/configuration.nix:8]: See above.

File: machines/stuart/configuration.nix [hunk 7]
```diff
@@ -7,7 +7,10 @@
   system.stateVersion = "22.05";
 
   lounge-rocks = {
-    oracle-aarch64.enable = true;
+    cloud-provider = {
+      enable = true;
+      oracle.enable = true;
+    };
```
Comment (pinpox) [inline machines/stuart/configuration.nix:13]: See above.

File: machines/woodpecker-server/configuration.nix [hunk 8]
```diff
@@ -17,13 +17,19 @@
...
+        enable = true;
+        interface = "enp1s0";
+        ipv6_address = "2a01:4f8:1c17:636f::";
+      };
     };
     nix-common.enable = true;
-    attic.server.enable = true;
+    attic.server = {
+      enable = true;
+      enableNginx = true;
```
Comment (pinpox) [inline machines/woodpecker-server/configuration.nix:31]: Not really a fan of putting nginx proxies into the modules itself, as it limits composability and doesn't allow other proxies. Also consider, we might not want to hard-code the port and listening address.

File: modules/attic-server/default.nix [hunk 9]
```diff
@@ -46,7 +47,7 @@ let cfg = config.lounge-rocks.attic.server; in
         # available options:
         # https://github.com/zhaofengli/attic/blob/main/server/src/config-template.toml
 
-        listen = "127.0.0.1:7373";
+        listen = mkIf cfg.enableNginx "127.0.0.1:7373";
```
Comment (pinpox) [inline modules/attic-server/default.nix:50]: See above

File: modules/cloud-provider/default.nix [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+  # Collects all files of a directory as a list of strings of paths
+  files = dir:
+    collect isString
+      (mapAttrsRecursive (path: type: concatStringsSep "/" path) (getDir dir));
+
+  # Filters out directories that don't end with .nix or are this file, also makes the strings absolute
+  validFiles = dir:
+    map (file: ./. + "/${file}") (filter (file: hasSuffix ".nix" file && file != "default.nix") (files dir));
+
+  cfg = config.lounge-rocks.cloud-provider;
```
Comment (pinpox) [inline modules/cloud-provider/default.nix:22]: Not a fan of these copy/pasted everywhere.

File: modules/cloud-provider/default.nix [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+in
+{
+  imports = validFiles ./. ++ [
```
Comment (pinpox) [inline modules/cloud-provider/default.nix:26]: I'd just list the files here or put them in their dependency parents.

File: modules/cloud-provider/default.nix [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    # It also reduces the size of the qcow2 image, which is good for backups.
+    services.fstrim = {
+      enable = true;
+      interval = "weekly";
+    };
+
+    # We want to standardize our partitions and bootloaders across all providers.
+    # See: https://github.com/lounge-rocks/the-lounge/issues/16
+    # Currently Hetzner and Proxmox are standardized.
+    disko.devices.disk.main = mkIf (cfg.hetzner.enable || cfg.proxmox.enable) {
```
Comment (pinpox) [inline modules/cloud-provider/default.nix:54]: If this is dependent on `hetzner.enable` or `proxmox.enable` why not just put it in their respective options?

File: modules/cloud-provider/hetzner.nix [hunk 13]
```diff
@@ -0,0 +1,50 @@
+{ config, lib, ... }:
+with lib;
+let cfg = config.lounge-rocks.cloud-provider.hetzner;
+
+in {
+
+  options.lounge-rocks.cloud-provider.hetzner = {
+    enable = mkEnableOption "activate hetzner";
```
Comment (pinpox) [inline modules/cloud-provider/hetzner.nix:8]: Verb get's autopopulated by `mkEnableOption` for docs
```suggestion
    enable = mkEnableOption "hetzner configuration";
```

File: modules/cloud-provider/hetzner.nix [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    interface = mkOption {
+      type = types.str;
+      default = "enp1s0";
+      description = "Interface to use";
+    };
+    ipv6_address = mkOption {
```
Comment (pinpox) [inline modules/cloud-provider/hetzner.nix:14]: Nitpick: we are mixing camelCase and snake_case all over the place.

File: modules/cloud-provider/hetzner.nix [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+      default = "NONE";
+      description = "IPv6 address of the server";
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    ### Networking ###
+    networking = {
+      interfaces.${cfg.interface} = {
```
Comment (pinpox) [inline modules/cloud-provider/hetzner.nix:25]:
```suggestion
      interfaces."${cfg.interface}" = {
```

File: modules/cloud-provider/netcup.nix [hunk 16]
```diff
@@ -1,14 +1,12 @@
...
+let cfg = config.lounge-rocks.cloud-provider.netcup;
 
 in {
 
-  imports = [ "${modulesPath}/profiles/qemu-guest.nix" ];
-
-  options.lounge-rocks.netcup-x86 = {
-    enable = mkEnableOption "activate netcup-x86";
+  options.lounge-rocks.cloud-provider.netcup = {
+    enable = mkEnableOption "activate netcup";
```
Comment (pinpox) [inline modules/cloud-provider/netcup.nix:9]:
```suggestion
  options.lounge-rocks.cloud-provider.netcup = {
    enable = mkEnableOption "netcup configuration";
```

File: modules/cloud-provider/oracle.nix [hunk 17]
```diff
@@ -19,7 +17,7 @@ in {
         efiInstallAsRemovable = true;
         device = "nodev";
       };
-      cleanTmpDir = true;
+      tmp.cleanOnBoot = true;
```
Comment (pinpox) [inline modules/cloud-provider/oracle.nix:20]: Is this something we might want on all machines/modules?

File: modules/cloud-provider/proxmox.nix [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+{ config, lib, modulesPath, ... }:
+with lib;
+let cfg = config.lounge-rocks.cloud-provider.proxmox;
+
+in {
+
+  options.lounge-rocks.cloud-provider.proxmox = {
+    enable = mkEnableOption "activate proxmox";
+  };
```
Comment (pinpox) [inline modules/cloud-provider/proxmox.nix:19]:
```suggestion
  options.lounge-rocks.cloud-provider.proxmox = {
    enable = mkEnableOption "proxmox configuration";
  };
```

File: modules/cloud-provider/proxmox.nix [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+  config = mkIf cfg.enable {
+
+    # If /dev/sda is not shown, you need to set your disk to VirtIO Block in the Proxmox GUI
+    lounge-rocks.cloud-provider.primaryDisk = "/dev/vda";
+
+    boot.initrd.availableKernelModules = [ "9p" "9pnet_virtio" "ata_piix" "uas" "uhci_hcd" "virtio_blk" "virtio_mmio" "virtio_net" "virtio_pci" "virtio_scsi" ];
```
Comment (pinpox) [inline modules/cloud-provider/proxmox.nix:26]:
```suggestion
    boot.initrd.availableKernelModules = [ 
      "9p"
      "9pnet_virtio"
      "ata_piix"
      "uas"
      "uhci_hcd"
      "virtio_blk"
      "virtio_mmio"
      "virtio_net"
      "virtio_pci"
      "virtio_scsi"
    ];
```

File: modules/cloud-provider/proxmox.nix [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+    boot.initrd.kernelModules = [ "virtio_balloon" "virtio_console" "virtio_rng" ];
+    boot.kernelModules = [ "kvm-intel" ];
+
```
Comment (pinpox) [inline modules/cloud-provider/proxmox.nix:29]:
```suggestion
```

File: modules/cloud-provider/proxmox.nix [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+# Processors: CPU type host
+# BIOS: OVMF (UEFI)
+# Machine: Default (i440fx)
+# SCSI Controller: VirtIO SCSI
+# Hard Disk: VirtIO Block
+
+# deploying a Proxmox system via nix-anywhere:
+# nix run github:numtide/nixos-anywhere -- --flake .#<host> root@<ip>
+
+{ config, lib, modulesPath, ... }:
```
Comment (pinpox) [inline modules/cloud-provider/proxmox.nix:11]:
```suggestion
{ config, lib, ... }:
```

File: modules/cloud-provider/proxmox.nix [hunk 22]
```diff
@@ -0,0 +1,31 @@
+# Proxmox PVE settings
+# Processors: CPU type host
+# BIOS: OVMF (UEFI)
+# Machine: Default (i440fx)
+# SCSI Controller: VirtIO SCSI
+# Hard Disk: VirtIO Block
+
```
Comment (pinpox) [inline modules/cloud-provider/proxmox.nix:7]: Does this still apply for us?

File: modules/cloud-provider/oracle.nix [hunk 23]
```diff
@@ -1,13 +1,11 @@
-{ config, pkgs, lib, modulesPath, ... }:
+{ config, pkgs, lib, ... }:
 
 with lib;
-let cfg = config.lounge-rocks.oracle-aarch64;
+let cfg = config.lounge-rocks.cloud-provider.oracle;
```
Comment (pinpox) [inline modules/cloud-provider/oracle.nix:4]: Formatting and indention seems off?

File: modules/cloud-provider/default.nix [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+    # This makes upgrading the size of the vDisk easier.
+    # TODO: can we do this through Disko?
+    # Haven't found something about this in the docs yet.
+    fileSystems."/".autoResize = true;
+    boot.growPartition = true;
+
+    # We want to standardize our partitions and bootloaders across all providers.
+    # See: https://github.com/lounge-rocks/the-lounge/issues/16
+    # Currently Hetzner and Proxmox are standardized.
+    boot.loader.grub = mkIf (cfg.hetzner.enable || cfg.proxmox.enable) {
```
Comment (pinpox) [inline modules/cloud-provider/default.nix:102]: Same for these checks. Might just want to have a full proxmox and hetzner-specific config

File: .drone.yml [hunk 25]
```diff
@@ -103,6 +103,16 @@ steps:
...
+  - name: Build attic-server
+    commands:
+      - nix build '.#nixosConfigurations.attic-server.config.system.build.toplevel'
+      - nix run .#s3uploader $(readlink result)
+    environment:
+      AWS_ACCESS_KEY_ID:
+        from_secret: aws_key
+      AWS_SECRET_ACCESS_KEY:
+        from_secret: aws_secret
+
```
Comment (pinpox) [inline .drone.yml:115]: Does a drone config even make sense at this point?

File: machines/attic-server/configuration.nix (hunk 4 above)
Comment (MayNiklas) [reply machines/attic-server/configuration.nix:17]: 👍 I will do that.

File: machines/attic-server/configuration.nix (hunk 2 above)
Comment (MayNiklas) [reply machines/attic-server/configuration.nix:4]: This won't work.
```log
 anonymous function at /nix/store/6jgwhd7idypzkv1bnw0krkgnl0sf864y-aa5lpybpjjpqpf6g3lj89hdj59y8y8bf-source/machines/attic-server/configuration.nix:1:1 called without required argument 'lib'

       at /nix/store/aa5lpybpjjpqpf6g3lj89hdj59y8y8bf-source/flake.nix:123:16:

          122|               sops-nix.nixosModules.sops
          123|               (import "${./.}/machines/${x}/configuration.nix" { inherit self; })
             |                ^
          124|             ];
```

File: machines/woodpecker-server/configuration.nix (hunk 8 above)
Comment (MayNiklas) [reply machines/woodpecker-server/configuration.nix:31]: That's a valid point!
I will remove nginx from the module once we deployed everything.
Current plan: reverse proxy won't be on the same server anyway - so we can completely remove it.

File: modules/cloud-provider/default.nix (hunk 12 above)
Comment (MayNiklas) [reply modules/cloud-provider/default.nix:54]: TLDR: our Netcup & Oracle servers need to be redeployed - after that, the check won't be necessary.
Every provider will share the same bootloader and partition config.
There isn't that much provider specific stuff after that (mainly kernel modules & boot options for the serial console).

File: .drone.yml (hunk 25 above)
Comment (MayNiklas) [reply .drone.yml:115]: We only have x86 runners on drone currently.
Yes: we will be able to delete .drone.yml once we have x86 runners and deployed attic in production.
Currently, drone still is responsible to filling our S3 cache.

File: modules/cloud-provider/default.nix (hunk 24 above)
Comment (MayNiklas) [reply modules/cloud-provider/default.nix:102]: See above.

File: modules/cloud-provider/proxmox.nix (hunk 22 above)
Comment (MayNiklas) [reply modules/cloud-provider/proxmox.nix:7]: Yes!
Those settings need to be set on the VM host.
The module works under the assumption everything is configured that way.

File: modules/cloud-provider/default.nix (hunk 10 above)
Comment (MayNiklas) [reply modules/cloud-provider/default.nix:22]: Will replace them!

File: machines/attic-server/configuration.nix (hunk 3 above)
Comment (MayNiklas) [reply machines/attic-server/configuration.nix:14]: How about we enable `cloud-provider.enable` within the provider specific configs?
This way we still have a general module for providers, that don't have any specific requirements.

Long time goal as I stated before: don't having much specifics.

**Comment by MayNiklas** [comment]
Thank you for the suggestions @pinpox !
I implemented some improvements addressing those.

Some stuff is a temporary solution, since it will be deleted once we are fully deployed / restructured (for example Nginx config for attic).

**Pull request merged**